---
title: "differential expression analysis of TCGA STAD using edgeR"
author: Camila Sinimbú Forte
date: 26/09/2024
output: html_notebook
---
```{r}
load("TCGA-STAD_clinical.rda")
```


```{r}
library("dplyr")
library("tidyverse")
library("edgeR")
library("EnhancedVolcano")
library("ggpubr")
library("ggplotify")
library("stringr")
library("org.Hs.eg.db")
library("purrr")
library("vtable")
```


```{r}
counts <- read.csv("TCGA-STAD_counts.txt", sep='\t', header=TRUE) #data.frame dos counts 

clinical_data <- tbl[c("submitter_id", "primary_diagnosis",
                       "gender", "vital_status", "age_at_index",
                       "treatments_pharmaceutical_treatment_or_therapy",
                       "treatments_radiation_treatment_or_therapy" )]

clinical_data <- na.omit(clinical_data)
```

```{r}
gene_description <- counts[,c("gene_id", "gene_name", "gene_type")] #lista com a descrição dos genes
rownames(counts) <- counts$gene_id #mudando o índice
counts <- counts[,!(names(counts) %in% c("gene_id", "gene_name", "gene_type"))] #xcluindo gene_name e gene_type
colnames(counts) <- gsub("[.]", "-", colnames(counts))
```

```{r}
case_samples_list <- clinical_data$submitter_id #lista dos submitter_id
case_counts <- counts[, colnames(counts) %in% case_samples_list] #counts com somente os presentes em clinical_data
table(clinical_data$primary_diagnosis)

clinical <- clinical_data[clinical_data$submitter_id %in% colnames(case_counts),]
```


```{r}
clinical_data$pharm <- -9 #criando coluna pharm
clinical_data$rad <- -9 #coluna rad

clinical_data[clinical_data$treatments_pharmaceutical_treatment_or_therapy == "no", ]$pharm <- 0 
clinical_data[clinical_data$treatments_pharmaceutical_treatment_or_therapy == "yes",]$pharm  <- 1 
clinical_data[clinical_data$treatments_pharmaceutical_treatment_or_therapy == "not reported", ]$pharm <- 2

clinical_data[clinical_data$treatments_radiation_treatment_or_therapy == "no", ]$rad <- 0
clinical_data[clinical_data$treatments_radiation_treatment_or_therapy == "yes", ]$rad <- 1
clinical_data[clinical_data$treatments_radiation_treatment_or_therapy == "not reported", ]$rad <- 2 #substituindo "não reportado" por 2

#excluindo as colunas antigas
clinical_data$treatments_pharmaceutical_treatment_or_therapy <- NULL 
clinical_data$treatments_radiation_treatment_or_therapy <- NULL
```



```{r}
#função para comparar os tipos
compare_deg <- function(case1, case2){
  
  #separa os submitter_id (samples) para cada subtipo de câncer
  tipo1.samples <- clinical_data[clinical_data$primary_diagnosis == case1, ]$submitter_id
  tipo2.samples <- clinical_data[clinical_data$primary_diagnosis == case2, ]$submitter_id

  #separa os tipos para os counts dos casos
  tipo1 <- case_counts[, colnames(case_counts) %in% tipo1.samples]
  tipo2 <- case_counts[, colnames(case_counts) %in% tipo2.samples]

  mat <- cbind(tipo1, tipo2) #merge dos casos

  #filtragem
  keep.genes <- filterByExpr(mat) #filterByExpr: mantém somente os counts com um número mínimo de amostras
  mat <- mat[keep.genes, ]
  mat <- cpm(mat) #normalização: counts por milhão

  tmat <- as.data.frame(t(mat)) #transposta
  tmat$samples <- rownames(tmat) #criando coluna samples com os indices de tmat
  tmat <- merge(tmat, clinical_data, by.x = "samples", by.y = "submitter_id") #merge de tmat com os dados clínicos a partir da coluna samples e submitter_id
  rownames(tmat) <- tmat$samples #os indices de tmat voltam a ser os samples
  tmat$samples <- NULL #a coluna sample é excluida

  exp <- t(tmat[, !(colnames(tmat) %in% colnames(clinical_data))])
  #exp são os counts
  
  #matriz de covariaveis
  cov <- tmat[, (colnames(tmat) %in% colnames(clinical_data))]

#construindo a matriz de design que descreve as condições(covariaveis) do modelo
  design_mat <- model.matrix(~cov$primary_diagnosis + cov$gender + cov$age_at_index + cov$pharm + cov$rad)

#classe de dados DGEList que recebe os counts, os genes e os grupos
#no nosso caso os subtipos de câncer de estômago
  dgList <- DGEList(counts=exp, genes=rownames(exp), group= cov$primary_diagnosis)

  #normalização
  #transformação de distribuição
  dgList <- calcNormFactors(dgList, method="TMM")
  #TMM é a média ponderada 
  
  #teste de expressão diferencial
  #estimateDisp: estimação das dispersões para cada gene
  
  #variabilidade geral em todo o genoma para este conjunto de dados
  dgList <- estimateDisp(dgList, design=design_mat)
 
  
  #glm: modelo linear generalizado
  #regressao linear - binomial negativa
  #1. ajustando o modelo para os counts e as covariaveis
  fit <- glmFit(dgList, design_mat)
  
  #2. teste de hipotese para ver se algum dos coeficientes (média de cada grupo) é diferente de zero
  lrt <- glmLRT(fit, coef = 2) 
  
  #principais genes diferencialmente expressos classificados pelo p-valor
  de_results <- topTags(lrt, adjust.method = "fdr", sort.by = "PValue", p.value = Inf, n = Inf)
  de_results <- de_results$table

  #Genome wide annotation for Human
  #anotação dos genes a partir dos transcritos
  #mapeando os ids, especificando as colunas symbol pelo ensembl
  de_results$symbol <- mapIds(org.Hs.eg.db, rownames(de_results)<-gsub("\\..*","", rownames(de_results)), 
                     column="SYMBOL", keytype = "ENSEMBL", multiVals="first")

  
  #criando uma coluna de keyvals que recebe uma cor para especificação de valor: regulado up, down e sem significância
   keyvals <- ifelse((de_results$logFC < 0 & de_results$FDR <= 0.05), 'royalblue', ifelse((de_results$logFC > 0 & de_results$FDR <= 0.05), 'gold', 'black'))
  keyvals[is.na(keyvals)] <- 'black'
  names(keyvals)[keyvals == 'gold'] <- 'Up-regulated'
  names(keyvals)[keyvals == 'black'] <- 'No significance'
  names(keyvals)[keyvals == 'royalblue'] <- 'Down-regulated'
  
  int_genes <- c("CFTR", "MUC2", "MUC4", "MUC6", "MUC5B", "MUC5AC")
  de_results$keyvals <- keyvals
  de_results.genes <- de_results[de_results$symbol %in% int_genes,]

  
  #contagem de ups, downs e notsig 
  sum <- summary(de <- decideTests(object = lrt, adjust.method = "fdr", p.value = 0.05, lfc = 0))
  sum <- data.frame(sum)
  rownames(sum)[1] <- sprintf("%s & %s", case1, case2)
  sum$down <- sum$Freq[1]
  sum$NotSig <- sum$Freq[2]
  sum$up <- sum$Freq[3]
  sum <- sum[-(2:3),-(1:3)]
  
  #table for summary
  #write.table(sum, file = sprintf("Summary_ups_downs_notsigs/summary_%s_%s.txt", case1, case2), quote = F, sep = "\t", row.names = F, eol = "\r")
  
  #table for deg results
  #write.table(de_results, file = sprintf("DEG_Results/DEG_%s_%s.txt", case1, case2), quote = F, sep = "\t", row.names = F, eol = "\r")
  
  #table for only cftr and mucins deg results
  #write.table(de_results.genes, file = sprintf("DEG_CFTR_MUCs/CFTR_MUC_%s_%s.txt", case1, case2), quote = F, sep = "\t", row.names = F, eol = "\r")
  
  return(c("done"))
  }
```


```{r}
diag = c("Carcinoma, diffuse type", "Adenocarcinoma, intestinal type", "Adenocarcinoma, NOS",
"Signet ring cell carcinoma", "Tubular adenocarcinoma", "Mucinous adenocarcinoma")

#combn(diag, 2, function(x) compare_deg(x[1], x[2]))
```





